### DISCUSSION TITLE
Blitz & Storybook

### DISCUSSION THREAD

**Discussion by juan-manuel-alberro** [blitz-js/blitz#2456] [Q&A]
Hi all,

Someone had the luck to install and run `Storybook` with `Blitz`?

By running `npx sb init` the installation is completed and working fine with the demo components but at the moment I try to import a component inside `app/core/components` there are plenty of issues with Webpack and module resolution.

I followed this example as well https://github.com/vercel/next.js/tree/canary/examples/with-storybook but without luck. 

Thanks.

**Comment by flybayer** [comment]
Does a relative import of `../core/components/XXXX` work?

**Reply by juan-manuel-alberro** [reply]
Nope, in my component, I'm using `Link` and `useRouter` from Blitz
```js
import { Link, useRouter } from "blitz"
```
 and there are some issues trying to import those files (including messages like 'fs' not found, for eg.).

If I import those files from Next
```js
import Link from "next/link"
import { useRouter } from "next/router"
```
It works, so I believe is something related to how Blitz was compiled?
This can be easily recreated by installing Storybook on top of a new Blitz app.

**Reply by flybayer** [reply]
That sounds like a babel issue. Is storybook configured to use `blitz/babel` preset? (fyi I haven't personally tried setting up storybook yet)

**Reply by juan-manuel-alberro** [reply]
Yes, I have the `.babelrc` inside the `.storybook` folder with

```js
{
  "presets": ["blitz/babel"],
  "plugins": []
}
```

And when I include for example the `Link` component from `blitz` in any story, there are several errors for example

```js
ERROR in ./node_modules/@blitzjs/config/dist/blitzjs-config.esm.js
Module not found: Error: Can't resolve 'fs' in '/home/user/Projects/test/node_modules/@blitzjs/config/dist'
 @ ./node_modules/@blitzjs/config/dist/blitzjs-config.esm.js 2:0-20 61:6-8
 @ ./node_modules/@blitzjs/core/dist/blitzjs-core.esm.js
 @ ./stories/Page.tsx
 @ ./stories/Page.stories.tsx
 @ ./stories sync ^\.(?:(?:^|\/|(?:(?:(?!(?:^|\/)\.).)*?)\/)(?!\.)(?=.)[^/]*?\.stories\.(js|jsx|ts|tsx))$
 @ ./.storybook/generated-stories-entry.js
```

**Reply by flybayer** [reply]
Ok thanks. Then likely the storybook webpack config needs fixed to exclude Blitz server files like this https://github.com/blitz-js/blitz/blob/canary/packages/core/src/with-blitz.ts#L42-L51 and this https://github.com/blitz-js/blitz/blob/canary/packages/core/src/with-blitz.ts#L86-L90.

I've used storybook very little, so idk, but hopefully there's a way we can easily customize the storybook config in Blitz apps.

**Reply by juan-manuel-alberro** [reply]
Thanks, Brandon for the reply.
By adding those exclusions Storybook builds fine without errors but if I include any component from Blitz the story disappear from the Storybook list (video attached).

I created a `DummyButton` with vanilla elements, and the `Accordion` using Blitz components (`Link`/`useRouter`), when I uncomment the Accordion code the file is removed from SB (there are no errors in the console and probably is related to the exclusions?)

https://user-images.githubusercontent.com/7114488/121373722-397fe780-c937-11eb-9f72-18fb2304bb64.mov

<details>
 <summary>storybook config (main.js)</summary>

```js
const excluded = [
  /[\\/]npm-which[\\/]/,
  /[\\/]cross-spawn[\\/]/,
  /@blitzjs[\\/]config/,
  /blitz[\\/]packages[\\/]config/,
  /blitz2[\\/]packages[\\/]config/,
]

module.exports = {
  stories: ["../app/**/*.stories*.@(ts|tsx)"], 
  addons: ["@storybook/addon-links", "@storybook/addon-essentials", "@storybook/addon-postcss"],
  webpackFinal: async (config, { configType }) => {
    config.module.rules.push({
      issuer: /(mutations|queries)(?!.*\.client)/,
      resource: /_resolvers/,
      use: { loader: "null-loader" },
    })

    excluded.forEach((excluded) => {
      config.module.rules.push({ test: excluded, use: { loader: "null-loader" } })
    })

    return config
  },
}

```
</details>

**Reply by flybayer** [reply]
Ok then that sounds like a babel config issue. Storybook must use the `blitz/babel` babel preset. Can you confirm it's using that?

**Reply by juan-manuel-alberro** [reply]
Yes, I have a `.babelrc` file inside `.storybook` with

```json
{
  "presets": ["blitz/babel"],
  "plugins": []
}
```

**Reply by flybayer** [reply]
Does it work with you import Link and useRouter directly from Next.js?

**Reply by juan-manuel-alberro** [reply]
Yes, if I import from

```js
import Link from "next/link"
import { useRouter } from "next/router"
```

it works with the same configuration.

**Reply by flybayer** [reply]
Ok can you open an Blitz issue for this with steps to reproduce?

**Reply by juan-manuel-alberro** [reply]
Done, it's a bit long but I tried to give all the info that I got 🙂  https://github.com/blitz-js/blitz/issues/2468
Thanks for your time.
